control-plane/p4RuntimeSerializer.cpp

```diff
@@ -976,6 +987,50 @@ class P4RuntimeSerializer {
         }
     }
 
+    void addControllerHeader(const IR::Type_Header* type) {
+        if (isHidden(type)) return;
+
+        auto name = controlPlaneName(type);
+        auto id = symbols.getId(P4RuntimeSymbolType::CONTROLLER_HEADER, name);
+        auto annotations = type->to<IR::IAnnotated>();
+
+        auto controllerAnnotation = type->getAnnotation("controller_header");
+        CHECK_NULL(controllerAnnotation);
+
+        if (controllerAnnotation->expr.size() != 1) {
+            ::error("@controller_header should be a string for declaration %1%", type);
+            return;
+        }
+        auto nameConstant = controllerAnnotation->expr[0]->to<IR::StringLiteral>();
+        if (nameConstant == nullptr) {
+            ::error("@controller_header should be a string for declaration %1%", type);
+            return;
+        }
+        auto controllerName = nameConstant->value;
+
+        auto header = p4Info->add_controller_packet_metadata();
+        header->mutable_preamble()->set_id(id);
+        // According to the p4info specification, we use the name specified in the annotation for
+        // the p4info preamble, not the P4 fully-qualified name.
+        header->mutable_preamble()->set_name(controllerName);
```

Comment:
So we're computing the id based on the P4 name, but we're assigning the name based on the annotation. That's different than any other P4Info object. I don't really see why it'd be a problem, though.